components/brave_ads/browser/ads_p3a.cc

```diff
@@ -0,0 +1,160 @@
+
+namespace brave_ads {
+namespace p3a {
+
+namespace {
+
+constexpr char kNewTabPageEventCountDictPref[] =
+    "brave.brave_ads.p3a.ntp_event_count";
+constexpr char kViewCountKey[] = "views";
+constexpr char kClickCountKey[] = "clicks";
+constexpr char kExpireTimeKey[] = "expiry_time";
+
+constexpr int kCountBuckets[] = {0, 1, 2, 3, 8, 12, 16};
+
+constexpr base::TimeDelta kHistogramExpiryTime = base::Days(7);
+constexpr base::TimeDelta kCountExpiryTime = base::Days(30);
+
+std::string GetHistogramName(
+    const std::string& creative_instance_id,
+    const ads::mojom::NewTabPageAdEventType event_type) {
+  std::string result = "creativeInstanceId." + creative_instance_id;
+  std::string metric_event_type;
+  switch (event_type) {
+    case ads::mojom::NewTabPageAdEventType::kServed:
+      NOTREACHED();
+      return {};
+    case ads::mojom::NewTabPageAdEventType::kViewed:
+      metric_event_type = "views";
+      break;
+    case ads::mojom::NewTabPageAdEventType::kClicked:
+      metric_event_type = "clicks";
+      break;
+  }
+  return base::StrCat({brave::kCreativeMetricPrefix, creative_instance_id, ".",
+                       metric_event_type});
+}
+
+void UpdateMetricCount(PrefService* local_state,
+                       const std::string& pref_key,
+                       const std::string& creative_instance_id,
+                       const ads::mojom::NewTabPageAdEventType event_type) {
+  DCHECK(local_state);
+
+  DictionaryPrefUpdate update(local_state, kNewTabPageEventCountDictPref);
+  int count = 0;
+
+  base::Value::Dict& update_dict = update->GetDict();
+
+  std::string full_pref_key =
+      base::StrCat({creative_instance_id, ".", pref_key});
+  std::string full_expiry_key =
+      base::StrCat({creative_instance_id, ".", kExpireTimeKey});
+
+  base::Value* current_value = update_dict.FindByDottedPath(full_pref_key);
+  if (current_value && current_value->is_int()) {
+    count = current_value->GetInt();
+  }
+
+  count++;
+  update_dict.SetByDottedPath(full_pref_key, base::Value(count));
+  base::Time new_expiry_time = base::Time::Now() + kCountExpiryTime;
+  update_dict.SetByDottedPath(full_expiry_key,
+                              base::TimeToValue(new_expiry_time));
+  p3a_utils::RecordToHistogramBucket(
+      GetHistogramName(creative_instance_id, event_type).c_str(), kCountBuckets,
+      count);
+}
+
+void DeleteOldCounts(PrefService* local_state) {
+  DCHECK(local_state);
+
+  const base::Value::Dict& dict =
+      local_state->GetDictionary(kNewTabPageEventCountDictPref)->GetDict();
+
+  std::vector<std::string> delete_list;
+
+  for (const auto [key, value] : dict) {
+    absl::optional<base::Time> expiry_time =
+        base::ValueToTime(value.GetDict().Find(kExpireTimeKey));
+    if (base::Time::Now() >= expiry_time.value_or(base::Time())) {
+      delete_list.push_back(key);
+    }
+  }
+
+  if (!delete_list.empty()) {
+    DictionaryPrefUpdate update(local_state, kNewTabPageEventCountDictPref);
```

Comment:
This can be moved outside of the for loop

Reply:
looks like it is outside the for loop?